### PR TITLE
[WIP] porting atimer.c to Rust

### PR DESCRIPTION
**Pull request by TheKK**
Open this PR to prevent duplicated work and early stage feedback.

My current plan for this PR could result in two phases:
1. port atimer.c directly
2. rewrite atimer.rs in Rusty way

The challenge here is to validate my changes and I'll do some investigation to see how to get it done.

### CODE REVIEW

File: rust_src/src/atimer.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+extern "C" {
+    static mut free_atimers: *mut Atimer;
+    static mut stopped_atimers: *mut Atimer;
+    static mut atimers: *mut Atimer;
+
+    static mut pending_signals: i32;
+}
+
+// Types of timers.
+// FIXME How large this enum is in C?
```
Comment (jeandudey) [inline rust_src/src/atimer.rs]: Maybe  in C you can change the enum to `#define`s and typedef the enum name to a fixed size type.
**Example:**
```c
enum ATIMER_TYPE {
   ATIMER_ABSOLUTE,
   ...
}
```

Changes to:
```c
typedef uint32_t ATIMER_TYPE;

#define ATIMER_ABSOLUTE 0
...
```

File: rust_src/src/atimer.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        while !p.is_null() && (p as *const Atimer != t) {
+            prev = p;
+
+            unsafe {
+                p = (*p).next;
+            }
+        }
+
+        if p as *const Atimer == t {
+            match prev.is_null() {
```
Comment (Wilfred) [inline rust_src/src/atimer.rs]: I think this would be clearer as a normal if statement.

File: rust_src/src/lib.rs [hunk 3]
```diff
@@ -39,6 +40,18 @@ pub use lists::Flistp;
 pub use floatfns::extract_float;
 pub use floatfns::fmod_float;
 
+// FIXME part of the used function is used inside mod atimer,
```
Comment (Wilfred) [inline rust_src/src/lib.rs]: Minor nit: 'part of these used functions are only used inside mod atimer'

File: src/atimer.c [hunk 4]
```diff
@@ -161,126 +146,126 @@ start_atimer (enum atimer_type type, struct timespec timestamp,
...
-  int i;
-  sigset_t oldset;
+/* void */
+/* cancel_atimer (struct atimer *timer) */
+/* { */
+/*   int i; */
+/*   sigset_t oldset; */
 
-  block_atimers (&oldset);
+/*   block_atimers (&oldset); */
```
Comment (Wilfred) [inline src/atimer.c]: Commented code can be removed.

**Comment by Wilfred** [comment]
I'm going to close this because it looks like you're no longer working on this. I'm very happy to still review / accept PRs if you fancy picking this up again.

**Pull request closed without merging**